Domains: Update instructions for mapping subdomains

leonardost opened the pull request:
#### Changes proposed in this Pull Request

Continuing the changes done in #51031 and #50668, this PR updates the **sub**domain mapping instructions that are presented to users. These instructions are shown in the thank you page, after the user buys the subdomain mapping, and in the domains management page, when a subdomain isn't mapped yet.

This PR presents CNAME or NS records mappings as recommended setups depending if the site is simple or AT, respectively. Alternative mapping setups (e.g. using A records) are shown as advanced setup options.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

Firstly the D59270-code patch should be applied.

Both simple and atomic sites should be tested.

1. In Calypso, go to Upgrades > Domains > Add a domain to this site
2. Select "Use a domain I own"
3. "Map your domain"
4. Insert a subdomain
5. Buy that subdomain mapping for your site
6. The "Thank you" page with further instructions to setup subdomain mapping will be shown.

The instructions should differ if the site is simple or atomic.

* For simple sites: The recommended setup should be to set a CNAME record pointing to the user's WPCOM domain, and there should be two advanced setup options:
  * The first should instruct the user to set NS records pointing to our name servers.
  * The second should show the default IP addresses (*.24 and *.25) to be set as A records;

<img width="582" alt="subdomain mapping simple 1_" src="https://user-images.githubusercontent.com/5324818/113192783-95dad280-9235-11eb-8acc-9a538528b276.png">

* For AT sites: The recommended setup should be to set NS records pointing to our name servers, and there should be two advanced setup options:
  * The first should instruct the user to set a CNAME record pointing to their WPCOM staging domain.
  * The second should show the AT site's IP addresses to be set as A records;

<img width="584" alt="subdomain mapping AT 1_" src="https://user-images.githubusercontent.com/5324818/113192798-9d01e080-9235-11eb-81e2-c099214a88e9.png">

CNAME mapping for AT sites should show a warning:

<img width="585" alt="subdomain mapping AT 2_" src="https://user-images.githubusercontent.com/5324818/113192811-a12dfe00-9235-11eb-8cbc-fc6bd09bd8c8.png">

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

leonardost committed:
Update instructions for mapping subdomains

leonardost committed:
Refactor mapping instruction components to reduce code duplication

leonardost committed:
Refactor mapping instruction components to improve readability

leonardost committed:
Add is atomic site check when showing (sub)domain mapping instructions

leonardost committed:
Retrieve WPCOM domain of a site mapping in checkout thank you page

That value is used to show the correct CNAME mapping instructions for
the subdomain that was bought.

leonardost committed:
Retrieve WPCOM domain of a site in domains management page

That value is used to show the correct CNAME mapping instructions for
a bought subdomain mapping.

leonardost committed:
Improve subdomain mapping instrucitons with CNAME

leonardost committed:
Rename methods in DomainMappingInstructions to better express their intent

leonardost committed:
Improve wording of subdomain mapping instructions with CNAMEs

a8ci18n commented on client/my-sites/domains/components/mapping-instructions/index.jsx:
```diff
@@ -30,13 +30,115 @@ class DomainMappingInstructions extends React.Component {
+		const { translate } = this.props;
+
+		const nameServerInstructionsMessage = translate(
+			'Please log into your domain name registrar account and {{strong}}update the name servers{{/strong}} of your domain to use the following values, as detailed in {{link}}these instructions{{/link}}:',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **17** times:
`translate( 'Please log into your account at your domain registrar and {{strong}}update the name servers{{/strong}} of your domain to use the following values, as detailed in {{link}}these instructions{{/link}}:' )` *ES Score: 9*

leonardost committed:
Map subdomain_part backend attribute to subdomainPart property

leonardost committed:
Show correct subdomain part in subdomain mapping instructions

leonardost committed:
Update domain fixtures to fix unit tests

hambai reviewed (commented):
Okay the code looks good - I tested it and it worked as described.

However - I'd argue that displaying the "update name servers" as recommended step is not really the best approach for Atomic sites. At least the way I understand it is we recommend to the customer to update the name servers for the **domain** (which means the root domain) when they map a subdomain to us. I have always assumed that when a customer maps a subdomain to us they want to keep their root domain (and other records) pointed somewhere else. So recommending this option as default seems a bit risky? I'd just keep it the same as simple sites (CNAME, NS, A records) for atomic too. But I might be wrong here - did this ordering come from HEs?

On that note - we also have a separate page for mapping subdomains in the support docs - https://wordpress.com/support/domains/map-subdomain/ - maybe we should link to it (and ask the guild to update it)?

Another thing I noticed (not related to this PR) - would it make sense to change the wording from "Advanced setup ..." to "Alternative setup ..."? Mostly because "advanced" means it requires more skills/knowledge but in reality each of those setup instructions is advanced at this point :)

leonardost commented:
Hi @hambai, thank you for the review!

I'm also not really sure about what is the recommended setup to map subdomains in AT sites, but P9HQHe-k3-p2 and @klimeryk believe that pointing the name servers should be the recommended way. Oh, I just noticed the text says `update the name servers of your **domain**`, but it should read `subdomain` 🤦  good catch! I'll change it now. It is possible to update the name servers of a subdomain independently of the root domain, right? (I'm new so I'm not sure of that either 😅)

The support page for mapping subdomains is being linked to in the instructions, but it will probably need to be updated because some instructions are stale now.

I guess the change from "Advanced" to "Alternative" makes sense, because they're really all advanced as you said 😆

klimeryk commented on client/lib/domains/get-wpcom-domain.js:
```diff
@@ -0,0 +1,28 @@
+	return domainObject.type === domainTypes.WPCOM;
+}
+
+function isWpcomStagingDomain( domainObject ) {
+	return domainObject.domain.endsWith( '.wpcomstaging.com' );
```
To be on the safe side and to simplify code a bit, you could write this as:

```suggestion
function isWpcomStagingDomain( { domain = '' } ) {
	return domain.endsWith( '.wpcomstaging.com' );
```

Even better if `{ domainName: domain = '' }` works, but I'm not sure that's valid JS 😅

leonardost replied:
Unfortunately it seems it is not valid JS indeed 😞 but I implemented the suggested modification 👍

klimeryk commented:
I see where the confusion came from - and it's warranted. The recommended approach is to update the _NS records_ for the _subdomain_. It's a subtle difference and it might be hard to explain that to a user. Maybe we could show a similar "raw format" as we do for the CNAME record? Something like:

```
blog.example.com IN NS ns1.wordpress.com
blog.example.com IN NS ns2.wordpress.com
blog.example.com IN NS ns3.wordpress.com
```

(if that's the correct format ;))

But we should definitely mention "NS records" in the explanation as well.

One more thing: it'd be good to add a notice like we do for the root domain Advanced case for the advanced cases here as well. Based on `P9HQHe-k3-p2` there are downsides to mapping via CNAME or A records. We should show them - at least when the site is Atomic.

hambai commented:
> I see where the confusion came from - and it's warranted. The recommended approach is to update the _NS records_ for the _subdomain_. It's a subtle difference and it might be hard to explain that to a user. Maybe we could show a similar "raw format" as we do for the CNAME record? Something like:
> 
> ```
> blog.example.com IN NS ns1.wordpress.com
> blog.example.com IN NS ns2.wordpress.com
> blog.example.com IN NS ns3.wordpress.com
> ```
> 
> (if that's the correct format ;))
> 
> But we should definitely mention "NS records" in the explanation as well.

Yeah, something like that should be good. I even wonder if we should try to explicitly explain somehow like "don't change your root domain name servers" in there.

> One more thing: it'd be good to add a notice like we do for the root domain Advanced case for the advanced cases here as well. Based on `P9HQHe-k3-p2` there are downsides to mapping via CNAME or A records. We should show them - at least when the site is Atomic.

Thank you for the link - that definitely explains it. It makes sense actually since the Atomic subdomain is wpcomstaging... I knew it I'm missing something :)

hambai commented on client/my-sites/domains/components/mapping-instructions/index.jsx:
```diff
@@ -30,13 +30,113 @@ class DomainMappingInstructions extends React.Component {
+		const { translate } = this.props;
+
+		const nameServerInstructionsMessage = translate(
+			'Please log into your domain name registrar account and {{strong}}update the name servers{{/strong}} of your domain to use the following values, as detailed in {{link}}these instructions{{/link}}:',
```
```suggestion
			'Please log into your domain name registrar account and {{strong}}add NS records{{/strong}} for you subdomain to use following values, as detailed in {{link}}these instructions{{/link}}:',
```

And then show the proper DNS record format - I believe the one @klimeryk provided is correct though the subdomain there should be FQDN ending with dot. So:
```
blog.example.com. IN NS ns1.wordpress.com
blog.example.com. IN NS ns2.wordpress.com
blog.example.com. IN NS ns3.wordpress.com
```

hambai replied:
And yeah - we should definitely ask the guild to update the instructions for mapping a subdomain ( https://wordpress.com/support/domains/map-subdomain/ ) to include the NS records setup instructions and link to it

leonardost committed:
Make small improvement when checking if a domain is WPCOM staging

leonardost committed:
Add NS records mapping instructions for subdomains

leonardost committed:
Add alert in instructions when mapping subdomains with CNAME records

leonardost committed:
Remove link from mapping subdomains with NS records instructions until docs are updated

leonardost commented:
@klimeryk and @hambai, thank you for the considerations!

I added the NS records in DNS record format in the instructions and added a warning when mapping subdomains to AT sites with CNAME records. I also updated the PR's description screenshots to reflect these updates.

@hambai which guild should we ask to update the subdomain mapping support docs? The domains guild?

leonardost committed:
Change "downgrade" to "change" in CNAME mapping instructions

olaseni commented on client/my-sites/domains/components/mapping-instructions/index.jsx:
```diff
@@ -30,13 +30,153 @@ class DomainMappingInstructions extends React.Component {
+			<FoldableFAQ
+				id="advanced-mapping-setup-ns-records"
+				key="advanced-mapping-setup-ns-records"
+				question={ this.props.translate( 'Advanced setup using NS records' ) }
```
I also realise that this may sound better read as 'Alternative'.

olaseni reviewed (approved):
I have a non-blocking suggestion, but this looks good.

leonardost committed:
Change "Advanced" by "Alternative"

a8ci18n commented on client/my-sites/domains/components/mapping-instructions/index.jsx:
```diff
@@ -48,10 +188,10 @@ class DomainMappingInstructions extends React.Component {
 		const { translate } = this.props;
 
-		const advancedSetupUsingARecordsTitle = translate( 'Advanced setup using root A records' );
+		const advancedSetupUsingARecordsTitle = translate( 'Alternative setup using root A records' );
```
:warning: This change will be queued for retranslation. We'll display the English text until then.

leonardost committed:
Update subdomain mapping with A records instructions wording

a8ci18n commented on client/my-sites/domains/components/mapping-instructions/index.jsx:
```diff
@@ -30,41 +30,191 @@ class DomainMappingInstructions extends React.Component {
+		const { domainName, translate } = this.props;
+
+		const advancedSetupUsingARecordsTitle = isSubdomain( domainName )
+			? translate( 'Alternative setup using root A records' )
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **17** times:
`translate( 'Advanced setup using root A records' )` *ES Score: 6*

a8ci18n commented on client/my-sites/domains/components/mapping-instructions/index.jsx:
```diff
@@ -30,41 +30,191 @@ class DomainMappingInstructions extends React.Component {
+			: translate( 'Alternative setup using A records' );
+		const aRecordMappingWarning = isSubdomain( domainName )
+			? translate(
+					'If you map a subdomain using A records rather than WordPress.com name servers, you will need to manage your subdomain’s DNS records yourself for any other services you are using with your subdomain, including email forwarding or email hosting (i.e. with Google Workspace or Titan)'
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **17** times:
`translate( 'If you map a domain using A records rather than WordPress.com name servers, you will need to manage your domain’s DNS records yourself for any other services you are using with your domain, including email forwarding or email hosting (i.e. with Google Workspace or Titan)' )` *ES Score: 8*

leonardost committed:
Invert A records instructions header wording between subdomain and root domain cases

a8ci18n commented on client/my-sites/domains/components/mapping-instructions/index.jsx:
```diff
@@ -30,41 +30,191 @@ class DomainMappingInstructions extends React.Component {
+
+		const advancedSetupUsingARecordsTitle = isSubdomain( domainName )
+			? translate( 'Alternative setup using A records' )
+			: translate( 'Alternative setup using root A records' );
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **17** times:
`translate( 'Advanced setup using root A records' )` *ES Score: 6*

klimeryk reviewed (approved):
Re-tested and LGTM! Awesome work 💖

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/5639193

Thank you @leonardost for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/5639193).
